File: app/frontend/src/types/dataset.ts
export type ColumnDType = "datetime" | "numeric" | "categorical" | "string";
export type Frequency = "infer" | "D" | "W" | "MS" | "M" | "H";

export interface ColumnInfo {
  name: string;
  dtype: ColumnDType;
  example_values: string[];
  null_fraction: number;
}

export interface DatasetPreview {
  id: string;
  filename: string;
  columns: ColumnInfo[];
  row_count: number;
  first_rows: Record<string, unknown>[];
}

export interface DateParts {
  year_col: string;
  month_col: string;
  day_col?: string | null;
}

export interface ColumnMapping {
  value_col: string;
  date_col?: string | null;
  date_parts?: DateParts | null;
  series_id_col?: string | null;
  freq?: Frequency;
}

export interface SeriesSummary {
  id: string;
  length: number;
  first_date: string | null;
  last_date: string | null;
  preview: number[];
}

export interface SeriesExtraction {
  dataset_id: string;
  inferred_freq: string | null;
  series: SeriesSummary[];
}

export interface DatasetSummary {
  id: string;
  filename: string;
  row_count: number;
  uploaded_at: string;
  size_bytes: number;
}

File: app/backend/timesfm_studio/schemas/dataset.py
"""Dataset ingestion and column-mapping schemas."""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator

ColumnDType = Literal["datetime", "numeric", "categorical", "string"]
Frequency = Literal["infer", "D", "W", "MS", "M", "H"]


class ColumnInfo(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str
    dtype: ColumnDType
    example_values: list[str] = Field(default_factory=list)
    null_fraction: float = 0.0


class DatasetPreview(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    filename: str
    columns: list[ColumnInfo]
    row_count: int
    first_rows: list[dict[str, Any]]


class DateParts(BaseModel):
    """Compose a date from two columns (e.g. Year + Month like the repo's data.csv)."""

    model_config = ConfigDict(extra="forbid")

    year_col: str
    month_col: str
    day_col: str | None = None


class ColumnMapping(BaseModel):
    """User's mapping of CSV columns to TimesFM's expected inputs."""

    model_config = ConfigDict(extra="forbid")

    value_col: str = Field(description="Numeric column holding the time series values.")
    date_col: str | None = Field(
        default=None,
        description="Single date column. Either this or date_parts must be provided.",
    )
    date_parts: DateParts | None = Field(
        default=None,
        description="Compose the date from Year+Month (+Day) columns.",
    )
    series_id_col: str | None = Field(
        default=None,
        description="Optional column that groups rows into separate series.",
    )
    freq: Frequency = "infer"

    @model_validator(mode="after")
    def check_date_source(self) -> "ColumnMapping":
        if self.date_col is None and self.date_parts is None:
            raise ValueError("Either date_col or date_parts must be provided.")
        if self.date_col is not None and self.date_parts is not None:
            raise ValueError("Provide either date_col or date_parts, not both.")
        return self


class SeriesSummary(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    length: int
    first_date: str | None = None
    last_date: str | None = None
    preview: list[float] = Field(default_factory=list)


class SeriesExtraction(BaseModel):
    model_config = ConfigDict(extra="forbid")

    dataset_id: str
    inferred_freq: str | None
    series: list[SeriesSummary]


class DatasetSummary(BaseModel):
    """Lightweight summary of an uploaded dataset (for listing)."""

    model_config = ConfigDict(extra="forbid")

    id: str
    filename: str
    row_count: int
    uploaded_at: str  # ISO 8601 from meta.json
    size_bytes: int
